backend/server/auth.js

```diff
@@ -0,0 +1,136 @@
+require('../db/models/models');
+const crypto = require('crypto');
+const User = require('../db/models/User');
+const Breed = require('../db/models/Breed');
+const Photo = require('../db/models/Photo');
+
+function hashPassword(password) {
+  // Change pbkdf2Sync to pbkdf2Sync. Pbkdf2 is normall asyncronous, so the hash was not being returned at the end of the function.
+  // Get rid of toString functions in favor of changing data type in User model. See comments in db/models/User.js
+  const salt = crypto.randomBytes(128).toString('base64');
+  const iterations = 10000;
+  const hash = crypto.pbkdf2Sync(password, salt, iterations, 64, 'sha512');
+
+  return {
+    salt: salt,
+    hash: hash,
+    iterations: iterations,
+  };
+}
+
+function isPasswordCorrect(savedHash, savedSalt, savedIterations, passwordAttempt) {
+  // TODO: Use pbkdf2Sync. See above. Also add keylen and digest params.
+  // TODO: Saved hash and result of pbkdf2 are buffers, so use Buffer.compare to check for equality. See: https://masteringjs.io/tutorials/node/buffer-compare
+  return (
+    savedHash ===
+    crypto.pbkdf2Synch(
+      passwordAttempt,
+      savedSalt,
+      savedIterations,
+      64,
+      'sha512',
+      (err, derivedKey) => {
+        if (err) {
+          console.log(err);
+        } else {
+          console.log(derivedKey);
+        }
+      }
+    )
+  );
+}
+
+module.exports = {
+  signup: function (req, res) {
+    console.log('body: ', req.body);
+    const encryption = hashPassword(req.body.hashed_password);
```

Comment:
Why are we calling a hashing function on an already hashed password?

Reply:
we hashed it once client side just for security, and hashing it again server side where we can store the salt and iterations

Reply:
Ok, makes sense